### ISSUE TITLE
Encode/Marshal

### ISSUE THREAD

**Issue by 89z** [paulmach/protoscan#5]
From another issue, you showed that you can decode bytes into an arbitrary object. However can you do the reverse? It would be useful to have some map or struct defined in this package, that you could build, then marshal. I say this because some servers expect a request body of protobuf encoded bytes.

**Comment by paulmach** [comment]
> I say this because some servers expect a request body of protobuf encoded bytes.

As stated in the readme. "Package protoscan is a low-level reader for protocol buffersencoded data in Golang. The main feature is the support for lazy/conditional decoding of fields."  This library is not for encoding. 

What are you trying to do? There are many ways to encode protobuf. Many ways to encode it in a performant way. If you're looking for low level support try https://pkg.go.dev/google.golang.org/protobuf/encoding/protowire but the code generators like https://github.com/gogo/protobuf are very fast.

**Comment by 89z** [comment]
> What are you trying to do?

I reject the idea of having to use a compiler for ProtoBuf. I think you should be able to Marshal and Unmarshal just like JSON. And really, that should be possible. If they had only added a single extra wiretype for messages, ProtoBuf would be more or less self describing. At any rate, I want a package that can decode arbitrary ProtoBuf (which this package seems to do), and can also encode some Map or Struct into ProtoBuf as well.

**Comment by paulmach** [comment]
> I think you should be able to Marshal and Unmarshal just like JSON.

But you can't. The format is not self describing enough. This is alluded to in this [comment](https://github.com/paulmach/protoscan/issues/4#issuecomment-1032072337). Is it a string, bytes, or a sub-type? you don't know. You might be able to figure it out for a specific input, like when reverse engineering a gRPC api, but not in general.

> If they had only added a single extra wiretype for messages, ProtoBuf would be more or less self describing

what's the wire type?  Why would you even want it vs using something else like JSON or BSON?

> At any rate, I want a package that can decode arbitrary ProtoBuf

why don't you use this plus some [protowire.EncodeTag](https://pkg.go.dev/google.golang.org/protobuf/encoding/protowire#EncodeTag)

```
package protoscan_test

import (
	"fmt"

	"github.com/paulmach/protoscan"
	"github.com/paulmach/protoscan/internal/testmsg"
	"google.golang.org/protobuf/proto"
)

type Field struct {
	Number   int
	WireType int
	Data     []byte
}

func Example_readIntoArray() {
	child := &testmsg.Child{
		Number:  proto.Int64(123),
		Numbers: []int64{1, 2, 3, -4, -5, -6, 7, 8},
		Grandchild: []*testmsg.Grandchild{
			{
				Number:  proto.Int64(111),
				Numbers: []int64{-1, 2, -3, 4, -5, 6, -7, 8},
			},
			{
				Number:  proto.Int64(-222),
				Numbers: []int64{1, -2, 3, -4, 5, -6, 7, -8},
			},
		},
		After: proto.Bool(true),
	}

	data, err := proto.Marshal(child)
	if err != nil {
		panic(err)
	}

	fields := []Field{}

	msg := protoscan.New(data)
	for msg.Next() {
		f := Field{
			Number:   msg.FieldNumber(),
			WireType: msg.WireType(),
		}
		start := msg.Index
		msg.Skip()
		end := msg.Index

		f.Data = msg.Data[start:end]
		fields = append(fields, f)
	}

	for _, f := range fields {
		fmt.Printf("%v\n", f)
	}

	// Output:
	// {100 0 [123]}
	// {200 2 [50 192 62 111 130 125 44 255 255 255 255 255 255 255 255 255 1 2 253 255 255 255 255 255 255 255 255 1 4 251 255 255 255 255 255 255 255 255 1 6 249 255 255 255 255 255 255 255 255 1 8]}
	// {200 2 [59 192 62 162 254 255 255 255 255 255 255 255 1 130 125 44 1 254 255 255 255 255 255 255 255 255 1 3 252 255 255 255 255 255 255 255 255 1 5 250 255 255 255 255 255 255 255 255 1 7 248 255 255 255 255 255 255 255 255 1]}
	// {300 0 [1]}
	// {300 0 [2]}
	// {300 0 [3]}
	// {300 0 [252 255 255 255 255 255 255 255 255 1]}
	// {300 0 [251 255 255 255 255 255 255 255 255 1]}
	// {300 0 [250 255 255 255 255 255 255 255 255 1]}
	// {300 0 [7]}
	// {300 0 [8]}
	// {3200 0 [1]}
}
```

I'll add a PR for this just to have it.

**Comment by 89z** [comment]
> Is it a string, bytes, or a sub-type? you don't know. You might be able to figure it out for a specific input, like when reverse engineering a gRPC api, but not in general.

You can generalize this. `string`, `bytes` and `message` all get passed as wire type 2 (length-delimited). But you **can** differ between strings and bytes [1], and not all bytes slices are valid messages. So you can run those tests, and if you still have overlap, then you can just parse the data as both types, and add both types to the output under the same field number, using the type as the discriminator. This wouldnt work with `protowire.Type` [2], as again it uses the same type for all three, so any implementation would need to create a new `string`, `bytes` and `message` type.

1. https://github.com/golang/go/blob/go1.17.6/src/net/http/sniff.go#L297-L309
2. https://godocs.io/google.golang.org/protobuf/encoding/protowire#Type

> what's the wire type?

https://developers.google.com/protocol-buffers/docs/encoding#structure

> Why would you even want it vs using something else like JSON or BSON?

If it was my choice, I would never use ProtoBuf ever again. Its an awful format. However some servers I deal with, require ProtoBuf request body, and return ProtoBuf response body.

> why don't you use this plus some `protowire.EncodeTag`

Thats a good idea, but in my case I wanted an implementation that treats the field name as first class citizen, so I ended up doing something like this:

~~~go
type Tag struct {
   protowire.Number
   Name string
}
~~~

Then I can use the `string` either as a type discriminator, or as the field name associated with the field number.

**Comment by paulmach** [comment]
> Thats a good idea, but in my case I wanted an implementation that treats the field name as first class citizen, so I ended up doing something like this:

Sounds like you know the code you need. This package will not do encoding.
